<strong> MAKE SURE YOU REMEMBER TO RUN THE CELL BELOW </strong>

In [1]:
from time import time
from math import floor
from math import pi 
import random

Note, we make use of the random library in this problem set.  To get a feel for what this does, run the following print statements

In [2]:
print random.sample(range(5), 5) # Produces 5 random values between 0 and 5
print random.sample(range(10),5) # Produces 5 random values between 0 and 10
print random.sample(range(10),10) # Produces 10 random values between 0 and 10

[1, 0, 2, 4, 3]
[3, 9, 7, 8, 6]
[2, 3, 0, 1, 8, 9, 5, 6, 7, 4]


<strong> Problem 1 </strong>: 6pts, 19.1.  I appreaciate we did not go over selection sort in lecture, but if you turn to page 137 of your book, you will find it explained with an outline of the algorithm provided.  Part of coding is also about translating algorithms into code, though use the skeleton below for help.  Again, you should only have to add at most 4 lines of code, though you can have less if you use multiple assignments.  This is done where I have placed comments.  

*Answer to Problem 1*: Using the guidance from the book, we see that we need to keep swapping values until we get the absolute smallest unsorted value that remains.  This is done as below.  

In [9]:
def sel_sort(vals):
    for jj in range(0,len(vals)):
        for kk in range(jj+1,len(vals)):
            if vals[kk]<vals[jj]: 
                    vals[jj],vals[kk] = vals[kk],vals[jj]
    return vals 

In [10]:
tstart = time()
sel_sort(random.sample(range(50000),10)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.0


In [11]:
tstart = time()
sel_sort(random.sample(range(50000),50)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.00400018692017


In [12]:
tstart = time()
sel_sort(random.sample(range(50000),100)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.000999927520752


In [13]:
tstart = time()
sel_sort(random.sample(range(50000),500)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.0199999809265


In [14]:
tstart = time()
sel_sort(random.sample(range(50000),1000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.085000038147


In [15]:
tstart = time()
sel_sort(random.sample(range(50000),5000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

1.61800003052


In [16]:
tstart = time()
sel_sort(random.sample(range(50000),10000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

6.19599986076


In [ ]:
tstart = time()
sel_sort(random.sample(range(50000),50000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

Okay, this never stopped running, so I gave up.  It takes a long time as you would expect for an $\mathcal{O}(n^{2})$ method.  We even see this empiraclly since for $n=5000$ elements, the time taken is $1.62$ seconds, while for $n=10000$ it is $6.2$ seconds.  Having increased the number of points by a factor of $2$, the time taken should go up by a factor of $4$, and sure enough $1.62\times 4=6.48$, which is not so far off from $6.2$ as we should expect.  

<strong> Problem 2 </strong>: 6pts, 19.2.  Since we went over merge_sort in class, I will go ahead and include the code here.  That being said, please provide an explanation for why merge_sort is so much faster than sel_sort.  Your answer should include a discussion of the difference between $\mathcal{O}(n^{2})$ and $\mathcal{O}(n\log_{2}(n))$ algorithms, and how the different choices of $n$ affect these estimates of worst case scenarios for operation counts.   

In [19]:
def merge_list(Alist,Blist):
    Clist = []
    while len(Alist)>0 and len(Blist)>0: # Run so long as neither list is empty.
        if Alist[0]<=Blist[0]: # Compare beginning of lists to each other.  
            nval = Alist.pop(0) # Pop removes an entry and returns it.  
        else:
            nval = Blist.pop(0)
        Clist.append(nval)
    if len(Alist)==0: # Decide which list is empty.
        Clist.extend(Blist)
    else:
        Clist.extend(Alist)
    return Clist

In [20]:
def merge_sort(vals):
    nvals = len(vals)
    if nvals==1:
        return vals        
    else:
        nhalf = int(floor(nvals/2)) 
        return merge_list(merge_sort(vals[0:nhalf]),merge_sort(vals[nhalf:]))        

*Answer to Problem 2*: Below, we present the timing results for `merge_sort()`.

In [21]:
tstart = time()
merge_sort(random.sample(range(50000),10)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.00399994850159


In [22]:
tstart = time()
merge_sort(random.sample(range(50000),50)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.00399994850159


In [23]:
tstart = time()
merge_sort(random.sample(range(50000),100)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.00400018692017


In [24]:
tstart = time()
merge_sort(random.sample(range(50000),500)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.00399994850159


In [26]:
tstart = time()
merge_sort(random.sample(range(50000),1000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.00799989700317


In [27]:
tstart = time()
merge_sort(random.sample(range(50000),5000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.0659999847412


In [28]:
tstart = time()
merge_sort(random.sample(range(50000),10000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

0.199000120163


In [29]:
tstart = time()
merge_sort(random.sample(range(50000),50000)) # You need to decide how many points to put in your list
tend = time()
ctime = tend-tstart
print ctime

1.22100019455


So, as you can see, there is just no comparison really in terms of performance, and `merge_sort()`, because it is an $\mathcal{O}(n\log_{2}n)$ method.  So for example, in going from $n=10^{4}$ to $n=5\times10^{4}$ elements, we expect to take

\begin{align*}
\frac{5\times10^{4}\log_{2}(5\times 10^{4})}{4\times10^{4}\log_{2}10} = 5 + \frac{5\log_{2}5}{4\log_{2}10} \approx 5.8737
\end{align*}

as long.  And we see that $.2\times5.8737 = 1.17$, which is very close to what we see above.  A similar comparison using `sel_sort()` says we should take up to $25$ times as long in going from $n=10^{4}$ to $n=5\times 10^{4}$ elements.  So the advantage here is clear. 

<strong> Problem 3 </strong>: 6pts, 25.4.  To do this, I would use the following skeleton code.  Note the tricky call

`rlen(vals[1:])`

What is the point of this?  Well, every time you call yourself, it means you will pass every entry of `vals` except for the first one.  Thus, in effect, everytime you call rlen from within, you cut off the first entry of `vals`, thus decreasing the length by one on each recursive call.  An explicit example would be if I had

`vals = [1,4,7,3]`

then 

`vals[1:] = [4,7,3]`

so if I start with 

`rlen(vals) = rlen([1,4,7,3])`

at the recursive function call, I will have 

`return rlen(vals[1:]) = return rlen([4,7,3])`

No `pop` or `del` necessary.  Neat huh!

*Answer to Problem 3*: So, we just need to keep updating by $1$ until nothing is left, whereupon, we return 0.

In [31]:
def rlen(vals):
    if vals == []:   
        return 0
    else:
        return 1 + rlen(vals[1:]) 

In [32]:
rlen([1,2,3,4]) # Test to see if your function works.

4

<strong> Problem 4 </strong>: 6pts, 25.5.  To do this, I would use the following skeleton code.  Note, I like having two functions here, a master calling function, `rmax`, and then one that does the recursion, `rstep`.  Use the trick from <strong> Problem 3 </strong> above.  

*Answer to Problem 4*: Basically, this comes down to comparing the first element of a list against the existing choice for the maximum value, which we call `mval`.  Then we keep cutting and comparing until we run out of list.  

In [39]:
def rstep(vals,mval):
    if vals==[]:
        return mval
    else:
        if vals[0]>mval: # You should make a comparison here
            mval = vals[0]
            return rstep(vals[1:],mval)
        else:
            return rstep(vals[1:],mval)
def rmax(vals):
    mval = vals[0]
    return rstep(vals[1:],mval)    

In [40]:
rmax([1,3,2,7,12,2,39,4]) # Test to see if your function works. 

39

<strong> Problem 5 </strong>: 6pts, 25.1.  Okay, I'm not sure what the author is going for on this one.  To do this in the most recursive way possible though is not completely obvious, at least to me.  But here goes.  First, let's compute up to the $n+1$ term of the Maclaurin series of $\cos(x)$ so that 

$$
\cos(x) \approx 1 - \frac{x^{2}}{2!} + \frac{x^{4}}{4!} + \cdots + (-1)^{n}\frac{x^{2n}}{(2n)!}
$$

Using what is called *Horner's Method*, we can use our recursive thinking to write the series as 

$$
\cos(x) \approx 1 - \frac{x^{2}}{2!} + \frac{x^{4}}{4!} + \cdots + (-1)^{n-1}\frac{x^{2(n-1)}}{(2(n-1))!}\left(1 - \frac{x^{2}}{2n(2n-1)} \right)
$$

See what I did there?  Using the recursive trick for each term we have discussed in class before, I factor the last two terms in the series as shown above.  I can keep doing this so that I get 

$$
\cos(x) = 1 - \frac{x^{2}}{2!} + \frac{x^{4}}{4!} + \cdots + (-1)^{n-2}\frac{x^{2(n-2)}}{(2(n-2))!}\left(1 - \frac{x^{2}}{(2n-2)(2n-3)}\left(1 - \frac{x^{2}}{2n(2n-1)} \right) \right)
$$


So to think about this recursively, I define a function, say $c_{0}(x)$, and set   

$$
c_{0}(x) = 1.
$$

To proceed, I define the next function $c_{1}(x)$ so that 

$$
c_{1}(x) = 1 - \frac{x^{2}}{(2n)(2n-1)} c_{0}(x) = 1 - \frac{x^{2}}{(2n)(2n-1)},
$$

since $c_{0}(x)=1$.  This motivates the general formula 

$$
c_{j}(x) = 1 - \frac{x^{2}}{(2(n-(j-1)))(2(n-(j-1))-1)}c_{j-1}(x), ~ j\geq 1.
$$

so that we have 

$$
\cos(x) \approx c_{n}(x)
$$

*Answer to Problem 5*: So, once we follow our rough 'rule' for recursion using the formula above, we just kind of fill in the blanks as in the following code.

In [52]:
from math import sqrt, pi

In [56]:
def rcos(jj,x,nval): 
    if jj==0:
        return 1. 
    else:
        return 1. - x**2/( (2.*(nval-(jj-1)))*(2.*(nval-(jj-1))-1) )*rcos(jj-1,x,nval)

Taking our expansion out to $n=10$, we get 

In [57]:
rcos(10,pi/4.,10) # Test your function.  You should be able to get something very close to 1/sqrt(2)

0.7071067811865476

And sure enough, we are about dead on with the fact that 

$$
\cos\left(\frac{\pi}{4}\right) = \frac{1}{\sqrt{2}}.
$$

since

In [58]:
print 1./sqrt(2.)

0.707106781187
